src/db.c

```diff
@@ -1085,13 +1086,21 @@ void scanCallback(void *privdata, const dictEntry *de) {
         char buf[MAX_LONG_DOUBLE_CHARS];
         int len = ld2string(buf, sizeof(buf), *(double *)dictGetVal(de), LD_STR_AUTO);
         key = sdsdup(keysds);
-        val = sdsnewlen(buf, len);
+        if (!data->no_values) {
+            val = sdsnewlen(buf, len);
+        }
     } else {
         serverPanic("Type not handled in SCAN callback.");
     }
 
     listAddNodeTail(keys, key);
-    if (val) listAddNodeTail(keys, val);
+    if (val) {
+        if (!data->no_values) {
+            listAddNodeTail(keys, val);
+        } else if (o && o->type == OBJ_ZSET) {
```

Comment:
This is to avoid freeing the `val = dictGetVal(de);` when `o->type == OBJ_HASH` if #12765 is merged.